### PR TITLE
Add UI and supporting code to set custom loadouts for rebel troops

### PR DESCRIPTION
**Pull request by jaj22**
### What type of PR is this.
1. [ ] Bug
2. [ ] Change
3. [X] Enhancement

### What have you changed and why?
Added a UI and supporting code to set partial custom loadouts for rebel troops. This is strictly vetted by unit type to avoid exploits: Only grenadiers can use grenade launchers, for example. Ammunition is only lightly configurable (choice of "normal" or "high").  If there aren't enough weapons of a type, they can't be selected although in some cases they may spawn with the "random" option. If you only want to force some aspects of a loadout and leave the rest random, this is also possible.

Optics button is a placeholder. Not sure whether to implement anything here. There's currently no vetting for item existence, so if you change modset within a saved game then you might need to reset the loadouts in the UI.

The button to access the UI is on the right in the commander menu. There was a free slot after I switched it to 2x5.

### Please specify which Issue this PR Resolves.
closes #XXXX

### Please verify the following and ensure all checks are completed.
1. [X] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

Needs a DS test but it uses similar mechanics to A3A_rebelGear so it should be fine.

### Is further testing or are further changes required?
1. [ ] No
2. [X] Yes (Please provide further detail below.)

### CODE REVIEW

**Comment by jaj22** [comment]
UI screenshot as requested:

**Review by killerswin2 — APPROVED** [review]
LGTM

File: A3A/addons/core/dialogs.hpp [hunk 1]
```diff
@@ -1774,13 +1774,13 @@ class commander_comm 		{
...
-			text = "";
+			text = "Customize loadouts"; //$STR_antistasi_dialogs_commander_comm_customLoadouts;
 			x = 0.482498 * safezoneW + safezoneX;
 			y = 0.365981 * safezoneH + safezoneY;
 			w = 0.175015 * safezoneW;
 			h = 0.0560125 * safezoneH;
-			tooltip = "";
-			action = "";
+			tooltip = "Customize loadouts for rebel AI troops"; //$STR_antistasi_dialogs_commander_comm_customLoadouts_tooltip;
+			action = "if (player == theBoss) then {closeDialog 0; createDialog ""A3A_customLoadoutsDialog""} else {[""Custom Loadouts"", ""Only commanders have access to this function""] call A3A_fnc_customHint}";
```
Comment (Tiny-DM) [inline A3A/addons/core/dialogs.hpp:1783]: // TODO: Localize
Or just localize it yourself.

**Review by ante185 — COMMENTED** [review]
This will to a degree invalidate #3083, but i suppose it's fine.

the optics selection should probably be removed for now, or set to always display random until there's any particular way we want it to work

**Review by ante185 — APPROVED** [review]
actually, lgtm

**Comment by jaj22** [comment]
> This will to a degree invalidate #3083, but i suppose it's fine.

To check I'm not missing something here, I don't see how it affects #3083, because the pistols are only used before a rifle is unlocked, and this UI only gives you any control once a rifle is unlocked?

Going to put the fix for Tiny's click-spam bug in here too, otherwise it'll collide.

**Comment by jaj22** [comment]
Added the likely fix for #3449. It's not easy to prove because it depends on specific network & thread timings.

File: A3A/addons/gui/functions/GUI/fn_customLoadoutsDialog.sqf [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            [_priWeapon, "Shotguns", _curClass] call _fnc_addItems;
+        };
+
+        private _secWeapon = _display displayCtrl A3A_IDC_CUSLOAD_SECWEAPON;
+        lbClear _secWeapon;
+        _secWeapon lbAdd "Random";
+        _secWeapon lbSetCurSel 0;
+        _curClass = _roleGear getOrDefault ["SecWeapon", "Random"];
+        if (_roleData#3 != "") then { [_secWeapon, _roleData#3, _curClass] call _fnc_addItems };
+        // disable if not valid?
```
Comment (rautamiekka) [inline A3A/addons/gui/functions/GUI/fn_customLoadoutsDialog.sqf:114]:
```suggestion
        // TODO disable if not valid?
```

**Review by rautamiekka — CHANGES_REQUESTED** [review]

File: A3A/addons/core/functions/Ammunition/fn_addPrimaryAndMags.sqf [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+if !(primaryWeapon _unit isEqualTo "") then {
+    if (_weapon == primaryWeapon _unit) exitWith {};
+    private _magazines = getArray (configFile / "CfgWeapons" / (primaryWeapon _unit) / "magazines");
+    {_unit removeMagazines _x} forEach _magazines;			// Broken, doesn't remove mags globally. Pain to fix.
+    _unit removeWeapon (primaryWeapon _unit);
+};
+
+private _categories = _weapon call A3A_fnc_equipmentClassToCategories;
+
+if ("GrenadeLaunchers" in _categories && {"Rifles" in _categories} ) then {
```
[the remark this answers is not in the document]
Comment (jaj22) [reply A3A/addons/core/functions/Ammunition/fn_addPrimaryAndMags.sqf:36]: It's wrong though. Should be comparing to -1

[the remark this answers is not in the document]
Comment (rautamiekka) [reply A3A/addons/core/functions/Ammunition/fn_addPrimaryAndMags.sqf:36]: Yeah, I only just now realized that mistake. I keep forgetting this thing doesn't return boolean. Fixed.

**Comment by ante185** [comment]
> > This will to a degree invalidate #3083, but i suppose it's fine.
> 
> To check I'm not missing something here, I don't see how it affects #3083, because the pistols are only used before a rifle is unlocked, and this UI only gives you any control once a rifle is unlocked?
> 
> Going to put the fix for Tiny's click-spam bug in here too, otherwise it'll collide.

Okay then i'm not sure, is this working like you intended it to? this is just the starting vanilla FIA arsenal.
image
Just following up to make sure, as the purpose of the random pistols were to round out the capability of the early game rebel ai, so it's not necessarily a benefit for the player to make them not equip them, AKA I'm unbothered by this bypassing that.

**Comment by Play3rgg** [comment]
> > > This will to a degree invalidate #3083, but i suppose it's fine.
> > 
> > 
> > To check I'm not missing something here, I don't see how it affects #3083, because the pistols are only used before a rifle is unlocked, and this UI only gives you any control once a rifle is unlocked?
> > Going to put the fix for Tiny's click-spam bug in here too, otherwise it'll collide.
> 
> Okay then i'm not sure, is this working like you intended it to? this is just the starting vanilla FIA arsenal. image Just following up to make sure, as the purpose of the random pistols were to round out the capability of the early game rebel ai, so it's not necessarily a benefit for the player to make them not equip them, AKA I'm unbothered by this bypassing that.

Please make the opacity higher on the drop down, the current UI hurts my brain

**Comment by jaj22** [comment]
>  Just following up to make sure, as the purpose of the random pistols were to round out the capability of the early game rebel ai, so it's not necessarily a benefit for the player to make them not equip them, AKA I'm unbothered by this bypassing that.

Ah, I forgot that I included shotguns & SMGs in the list. Do you think any of these options is preferable?:
1. Include pistols if there are no rifles.
2. Include pistols regardless.
3. Don't show anything until there are rifles.

**Comment by jaj22** [comment]
> Please make the opacity higher on the drop down, the current UI hurts my brain

I remember trying that with the startup parameters and it was actually worse (more background/foreground confusion).

**Comment by jaj22** [comment]
The idea of 3 would be to still show SMG and shotgun choices, but only once you've unlocked a rifle. Doesn't make sense in the case where players unlock a better shotgun first though. Will probably leave it as is: In practice I doubt many people would be forcing a single-barrel shotgun over a random handgun choice.

File: A3A/addons/core/functions/Ammunition/fn_addPrimaryAndMags.sqf (hunk 3 above)
[the remark this answers is not in the document]
Comment (jaj22) [reply A3A/addons/core/functions/Ammunition/fn_addPrimaryAndMags.sqf:36]: Still wrong. It's an AND, while findAny can only do an OR.

**Review by killerswin2 — APPROVED** [review]

**Pull request merged**